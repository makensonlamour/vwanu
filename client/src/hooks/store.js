import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../features/auth/authSlice";
import alertReducer from "../features/alert/alertSlice";
import userReducer from "../features/user/userSlice";
import { apiSlice } from "../features/api/apiSlice";

export const store = configureStore({
  reducer: {
    user: userReducer,
    alert: alertReducer,
    auth: authReducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiSlice.middleware),
});
