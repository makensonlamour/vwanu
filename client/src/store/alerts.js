import { createSlice } from "@reduxjs/toolkit";
const initialState = {
  list: [],
};
const Alert = createSlice({
  name: "alerts",
  initialState,
  reducers: {
    setAlert: (state, action) => {
      state.list.push(action.payload);
    },
    removeAlert: (state, action) => {
      const firstIndex = state.list.findIndex((alert) => alert.id === action.payload);

      state.list.splice(firstIndex, 1);
    },
  },
});

export default Alert.reducer;
export const { removeAlert, setAlert } = Alert.actions;
export const getAlerts = (state) => state.alerts;
