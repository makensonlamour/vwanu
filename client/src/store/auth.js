import * as action from "./api";
import env from "../config";
import { createSlice } from "@reduxjs/toolkit";
import jwtDecode from "jwt-decode";
import { saveToken, deleteToken } from "../middleware/api";

const url = env.endpoints;
//const { EXPECTED_HEADER } = env;

const initialState = {
  loading: false,
  data: null,
  lastFetch: null,
  token: null,
  error: null,
};

export const Auth = createSlice({
  name: "auth",
  initialState,
  reducers: {
    loginRequested: (state) => {
      state.loading = true;
    },
    loginSucceed: (state, action) => {
      state.loading = false;
      state.token = action.payload.data.token;
      saveToken(action.payload.data.token);
      //state.data = action.payload.data.user;
      state.data = jwtDecode(action.payload.data.token);
      state.lastFetch = Date.now;
    },

    LoginFailed: (state, action) => {
      state.loading = false;
      state.token = null;
      state.error = action.payload;
      state.data = null;
      deleteToken();
    },
    setUser: (state, action) => {
      state.loading = false;
      state.token = action.payload;
      saveToken(action.payload);
      state.data = jwtDecode(action.payload);
    },
    LogOut: (state) => {
      state.loading = false;
      state.token = null;
      state.data = null;
      state.error = null;
      deleteToken();
      window.location.replace("/login");
    },
  },
});

export const Login = (credentials) => (dispatch, getState) => {
  const { data, loading } = getState();
  if (data != null || loading) return;
  dispatch(
    action.apiCallBegan({
      url: url.LOGIN,
      data: credentials,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const createUser = (newUserData) => (dispatch) => {
  dispatch(
    action.apiCallBegan({
      url: url.REGISTER,
      data: newUserData,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const forgotPassword = (newUserData) => (dispatch) => {
  dispatch(
    action.apiCallBegan({
      url: url.FORGOT_PASSWORD,
      data: newUserData,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const resetPassword = (newUserData) => (dispatch) => {
  dispatch(
    action.apiCallBegan({
      url: url.RESET_PASSWORD,
      data: newUserData,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const { setUser } = Auth.actions;
export const logout = () => (dispatch) => dispatch(Auth.actions.LogOut());
export const getCurrentUser = (state) => state.authentication;
export const logged = Auth.actions.loginSucceed.type;
export default Auth.reducer;
