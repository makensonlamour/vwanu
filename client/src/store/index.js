import { configureStore } from "@reduxjs/toolkit";
import thunkMiddleware from "redux-thunk";
import reducer from "./reducer";
import api from "../middleware/api";
import alerts from "../middleware/alerts";

export const store = configureStore({
  reducer,
  middleware: [thunkMiddleware, api, alerts],
});
