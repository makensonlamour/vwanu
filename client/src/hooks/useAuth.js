import { useState, useEffect } from "react";
// core components
import client from "../features/feathers";
import useAuthContext from "./useAuthContext";
import { register } from "../features/auth/authSlice";

const useAuth = () => {
  const [error, setError] = useState(null);
  const [isCancelled, setCancel] = useState(false);
  const [isLoading, setLoading] = useState(false);
  const { dispatch, Types } = useAuthContext();

  const signup = async ({ firstName, lastName, email, password, passwordConfirmation, termOfUse }) => {
    setError(null);
    setLoading(true);
    try {
      //signup the user
      const res = await register({ firstName, lastName, email, password, passwordConfirmation, termOfUse });

      //if success try login the user
      if (res?.data) {
        const responseLogin = await client.authenticate({
          strategy: "local",
          email,
          password,
        });

        // dispatch the user to the whole application
        if (responseLogin?.data) {
          dispatch({ type: Types.USER_LOGGED_IN, payload: responseLogin?.data });
        }
      }

      //dispatch({ type: Types.USER_LOGGED_IN, payload: user });
      if (!isCancelled) {
        setLoading(false);
        setError(null);
      }
    } catch (error) {
      console.error(error);
      if (!isCancelled) {
        setError(error.message);
        setLoading(false);
      }
    }
  };
  const login = async ({ email, password }) => {
    setLoading(true);
    setError(null);
    try {
      // try to sign in with provided data
      const res = await client.authenticate({
        strategy: "local",
        email,
        password,
      });
      console.log(res.User);
      //   dispatch user to the whole app if successful
      dispatch({ type: Types.USER_LOGGED_IN, payload: res.User });

      if (!isCancelled) {
        setLoading(false);
        setError(null);
      }
    } catch (error) {
      if (!isCancelled) {
        setLoading(false);
        setError(error.message);
      }
    }
  };

  const logout = async () => {
    setLoading(true);
    try {
      // get the current user ,
      // stop refreshing token
      // logout the user and
      //   const { uid } = auth.currentUser;
      if (!isCancelled) {
        setLoading(false);
        setError(null);
      }
      // Let the whole app knows the user is out
      dispatch({ type: Types.USER_LOGGED_OUT });
      window.location.reload();
    } catch (error) {
      if (!isCancelled) {
        setError(error.message);
        setLoading(false);
      }
    }
  };
  useEffect(() => {
    // clean up function
    return () => {
      setCancel(true);
    };
  }, []);

  return { logout, signup, login, isLoading, error };
};

export default useAuth;
