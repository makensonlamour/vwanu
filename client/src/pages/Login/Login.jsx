import React from "react";
import * as Yup from "yup";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import routesPath from "../../routesPath";
import { getAlerts } from "../../store/alerts";

// Core components
import Alerts from "../../components/common/Alerts";
import { Field, Form, Checkbox, Submit } from "../../components/form";
import { Login } from "../../store/auth";
import logo from "../../assets/images/Asset_3.png";
import WelcomeImg from "../../assets/images/welcome.png";
import { BsFacebook, BsTwitter } from "react-icons/bs";
import { FaGooglePlus } from "react-icons/fa";

const ValidationSchema = Yup.object().shape({
  email: Yup.string().required().min(6).email().label("Email"),
  password: Yup.string().required().min(8).label("Password"),
});

const LoginScreen = () => {
  const dispatch = useDispatch();
  const alerts = useSelector(getAlerts);
  const handleLogin = (credentials) => dispatch(Login(credentials));
  return (
    <>
      <div className="">
        <div className="grid grid-cols-1 lg:mb-48 lg:grid-cols-2 lg:gap-4 xl:mx-36 2xl:mx-96">
          <div className="lg:justify-center">
            <div className="bg-gradient-to-tr from-blue-400 to-secondary  h-96 z-0 lg:rounded-t-3xl lg:h-[90%] lg:m-6">
              <p className="text-center pt-14 text-base-100 text-2xl py-5">
                <img className="w-1/3 m-auto" src={logo} alt="logo_vwanu" />
              </p>
              <p className="text-5xl text-v-yellow-dark font-bold py-8 text-center">Welcome</p>
              <p className="hidden lg:block text-base-100 text-xl font-semibold text-center pb-2">Share your Voice and Change</p>
              <p className="hidden lg:block text-base-100 text-xl font-semibold text-center">The Haitian Community!</p>
            </div>
            <div className="hidden lg:block bg-v-yellow rounded-b-3xl rounded-tr-[130px] h-[50%] -mt-[9rem] m-6"></div>
            <div className="hidden lg:block mr-20 z-50 -mt-[375px]">
              {" "}
              <img className="object-fit" src={WelcomeImg} alt="young_people" />
            </div>
          </div>

          <div className="place-items-center bg-base-100 shadow-t-2xl rounded-t-[30px] px-8 -mt-28 z-10 lg:mt-6 lg:mx-10 md:mx-36 2xl:px-24">
            <div className="hidden lg:block lg:place-content-center lg:my-8">
              <p className="text-center mt-4">
                <span className="text-md text-primary font-semibold md:text-xl">Not a member ?</span>{" "}
                <Link
                  data-testid="registerBtn"
                  to={routesPath.REGISTER}
                  className="btn btn-secondary btn-sm rounded-full normal-case hover:btn-primary hover:text-base-100 md:px-8 ml-1 text-base-100"
                >
                  Register
                </Link>
              </p>
            </div>
            <Form
              validationSchema={ValidationSchema}
              initialValues={{
                email: "",
                password: "",
              }}
              onSubmit={handleLogin}
              className="mt-4 lg:shadow-2xl lg:rounded-t-3xl"
            >
              <h1 className="card-title text-primary font-bold text-xl lg:text-2xl">Sign in to Vwanu</h1>
              <Alerts className="bg-red-200" alerts={alerts} />
              <Field
                required
                autoCapitalize="none"
                placeholder="Email"
                name="email"
                type="email"
                autoComplete="new-email"
                className="mt-1 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-error autofill:text-secondary autofill:bg-blue-200"
                testId="email-error-message"
              />
              <Field
                required
                autoCapitalize="none"
                autoCorrect="false"
                placeholder="Password"
                name="password"
                autoComplete="new-email"
                className="mt-1 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-error autofill:text-secondary autofill:bg-blue-200"
                testId="password-error-message"
                showPassword={true}
              />
              <Checkbox
                name="remmberMe"
                label="Remember me"
                link={
                  <Link className="text-primary font-bold hover:text-secondary" to={routesPath.FORGOT_PASSWORD}>
                    Forgot password
                  </Link>
                }
                className=""
              />

              <Submit data-testid="loginBtn" className="rounded-full text-md btn-md" title="Login" />
              <div className="mt-9 text-center">
                <span className="text-blue-600 text-center inline text-md md:text-lg">
                  {` Or sign in with `}
                  <Link className="text-blue-500 ml-2" to={"#"}>
                    <BsFacebook className="text-2xl inline mx-1" />
                  </Link>
                  <Link className="text-sky-400" to={"#"}>
                    <BsTwitter className="text-2xl inline mx-1" />
                  </Link>
                  <Link className="text-red-600" to={"#"}>
                    <FaGooglePlus className="text-2xl inline mx-1" />
                  </Link>
                </span>
              </div>
              <div className="mt-8 lg:hidden">
                <p className="text-center mt-4">
                  <span className="text-md text-primary font-semibold md:text-xl">Not a member ?</span>{" "}
                  <Link
                    to={routesPath.REGISTER}
                    className="btn btn-secondary btn-sm rounded-full text-base-100 normal-case md:px-8 ml-1 hover:bg-primary"
                  >
                    Register
                  </Link>
                </p>
              </div>
            </Form>
          </div>
        </div>
        <div className="bg-gray-100 shadow-lg rounded-t-lg lg:mx-6 px-6 py-6">
          <span className="text-secondary font-semibold">Language:</span>
        </div>
      </div>
    </>
  );
};

export default LoginScreen;
