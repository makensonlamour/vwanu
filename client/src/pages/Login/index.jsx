import React from "react";
import * as Yup from "yup";
import { Link } from "react-router-dom";
import { useDispatch } from "react-redux";
import routesPath from "../../routesPath";

// Core components
import { Field, Form, Submit } from "../../components/form";
import { Login } from "../../store/auth";

const ValidationSchema = Yup.object().shape({
  email: Yup.string().required().min(6).label("Email"),
  password: Yup.string().required().min(8).label("password")
});

const LoginScreen = () => {
  const dispatch = useDispatch();
  const handleLogin = (credentials) => dispatch(Login(credentials));
  return (
    <>
      <div className="">
        <div className="grid grid-cols-2 gap-4">
          <div className="justify-center">
            <div className="bg-yellow-400 rounded-br-[300px] h-screen">
              <div className="bg-blue-600 h-4/6 rounded-br-full rounded-bl-[5000px]">
                <p className="text-center text-white text-2xl py-5">Logo VWANU</p>
                <p className="text-5xl text-yellow-500 font-bold py-10 text-center">Welcome</p>
                <p className="text-white text-xl font-semibold text-center">Share your Voice and Change</p>
                <p className="text-white text-xl font-semibold text-center">The Haitian Community!</p>
              </div>
            </div>
          </div>

          <div className="place-items-center mx-20">
            <div className="place-content-end my-8">
              <p className="text-right">
                <span className="text-orange-500 text-lg font-semibold">Not a member ?</span>{" "}
                <Link to={routesPath.REGISTER} className="btn btn-sm btn-primary px-8 ml-1 rounded-full text-base-200 normal-case">
                  Register
                </Link>
              </p>
            </div>
            <Form
              validationSchema={ValidationSchema}
              initialValues={{
                email: "",
                password: ""
              }}
              onSubmit={handleLogin}
              className="shadow-lg rounded-3xl"
            >
              <h1 className="card-title text-orange-500">Sign in to Vwanu</h1>
              <Field
                autoCapitalize="none"
                placeholder="Email"
                name="email"
                type="email"
                autoComplete="new-email"
                className="bg-blue-200 text-blue-500 font-semibold rounded-full px-6 input-primary border-none"
              />
              <Field
                autoCapitalize="none"
                autoCorrect="false"
                placeholder="Password"
                name="password"
                type="password"
                autoComplete="new-email"
                className="bg-blue-200 text-blue-500 font-semibold rounded-full px-6 input-primary border-none"
              />

              <Submit className="rounded-full text-base-200 text-md" title="Login" />
            </Form>
          </div>
        </div>
      </div>
    </>
  );
};

export default LoginScreen;
