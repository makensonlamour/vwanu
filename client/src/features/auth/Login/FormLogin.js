import React, { useEffect } from "react";
import * as Yup from "yup";
import { alertService } from "../../../components/common/Alert/Services";
import { Alert } from "../../../components/common/Alert";
import { Field, Form, Checkbox, Submit } from "../../../components/form";
import { BsFacebook, BsTwitter } from "react-icons/bs";
import { FaGooglePlus } from "react-icons/fa";
import Loader from "../../../components/common/Loader";
import { Link } from "react-router-dom";
import routesPath from "../../../routesPath";
// import { login } from "../authSlice";
import useAuth from "../../../hooks/useAuth";

const FormLogin = () => {
  const { isLoading, error, login } = useAuth();
  console.log(error);

  const ValidationSchema = Yup.object().shape({
    email: Yup.string().required().min(6).email().label("Email"),
    password: Yup.string().required().min(8).label("Password"),
  });

  const handleLogin = async (credentials) => {
    await login(credentials);
  };

  useEffect(() => {
    if (error !== null) alertService.error(error, { autoClose: true });
  }, [error]);

  return (
    <>
      <Form
        validationSchema={ValidationSchema}
        initialValues={{
          email: "",
          password: "",
        }}
        onSubmit={handleLogin}
        className="px-2 mt-4 lg:mt-4 lg:mx-2 xl:mx-14 3xl:mx-64"
      >
        <h1 className="card-title pb-2 text-primary font-bold text-xl lg:text-2xl">Sign in to Vwanu</h1>
        <Alert />
        <Field
          required
          autoCapitalize="none"
          placeholder="Email"
          name="email"
          type="email"
          autoComplete="new-email"
          containerClassName="my-3"
          className="mt-2 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-red-500 autofill:text-secondary autofill:bg-blue-200"
          testId="email-error-message"
        />
        <Field
          required
          autoCapitalize="none"
          autoCorrect="false"
          placeholder="Password"
          name="password"
          autoComplete="new-email"
          containerClassName="my-3"
          className="mt-2 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-red-500 autofill:text-secondary autofill:bg-blue-200"
          testId="password-error-message"
          showPassword={true}
        />
        <Checkbox
          name="rememberMe"
          label="Remember me"
          link={
            <Link className="font-semibold hover:text-primary" to={routesPath.FORGOT_PASSWORD}>
              Forgot password
            </Link>
          }
          className=""
        />
        <Submit
          disabled={isLoading ? true : false}
          data-testid="loginBtn"
          className="mt-2 rounded-full text-md btn-md"
          title={
            isLoading ? (
              <div className="flex justify-center">
                <Loader color="black" />
              </div>
            ) : (
              "Login"
            )
          }
        />
        <div className="mt-9 text-center">
          <span className="text-blue-600 text-center inline text-md md:text-lg">
            {` Or sign in with `}
            <Link className="text-blue-500 ml-2" to={"#"}>
              <BsFacebook className="text-2xl inline mx-1" />
            </Link>
            <Link className="text-sky-400" to={"#"}>
              <BsTwitter className="text-2xl inline mx-1" />
            </Link>
            <Link className="text-red-600" to={"/api/oauth/google"}>
              <FaGooglePlus className="text-2xl inline mx-1" />
            </Link>
          </span>
        </div>
        <div className="mt-8 lg:hidden">
          <p className="text-center mt-4">
            <span className="text-md text-primary font-semibold md:text-xl">Not a member ?</span>{" "}
            <Link
              to={routesPath.REGISTER}
              className="btn btn-secondary btn-sm rounded-full text-base-100 normal-case md:px-8 ml-1 hover:bg-primary"
            >
              Register
            </Link>
          </p>
        </div>
      </Form>
    </>
  );
};

export default FormLogin;
