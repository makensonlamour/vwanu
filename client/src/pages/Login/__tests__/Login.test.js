import React from "react";
import { render as rtlRender, screen, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { act as testAct, create } from "react-test-renderer";
import { act as domAct } from "react-dom/test-utils";
import { store } from "../../../hooks/store";
import Login from "../Login";
import { Provider } from "react-redux";

afterEach(cleanup);

const render = (component) => rtlRender(<Provider store={store}>{component}</Provider>);

//Testing Login form
//Test Dom Elements
describe("Test Dom Elements", () => {
  it("match snapshot", () => {
    let tree;
    domAct(() => {
      testAct(() => {
        tree = create(
          <Provider store={store}>
            <Router>
              <Login />
            </Router>
          </Provider>
        );
      });
    });
    expect(tree).toMatchSnapshot();
  });
});

it("should have a login form", () => {
  render(
    <Router>
      <Login />
    </Router>
  );

  const loginForm = screen.getByRole("form");
  expect(loginForm).toBeInTheDocument();
});

it("the login form should have an email input", () => {
  render(
    <Router>
      <Login />
    </Router>
  );

  const emailInput = screen.getByPlaceholderText("Email");
  expect(emailInput).toBeInTheDocument();
});

it("the login form should have a password input", () => {
  render(
    <Router>
      <Login />
    </Router>
  );

  const passwordInput = screen.getByPlaceholderText("Password");
  expect(passwordInput).toBeInTheDocument();
});

it("the login form should have a Login button", () => {
  render(
    <Router>
      <Login />
    </Router>
  );

  const submitButton = screen.getByRole("button", { name: "Login" });
  expect(submitButton).toBeInTheDocument();
});

it("the submit button should have the text Login", () => {
  render(
    <Router>
      <Login />
    </Router>
  );

  const submitButton = screen.getByRole("button", { name: "Login" });
  expect(submitButton).toHaveTextContent("Login");
});

it("the login page should have a register link", () => {
  render(
    <Router>
      <Login />
    </Router>
  );

  const registerLink = screen.getByTestId("registerBtn");
  expect(registerLink).toHaveTextContent("Register");
});

//Test require fields
describe("Test require fields", () => {
  it("the email input should be a required fields", () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    expect(emailInput).toBeInTheDocument();
    expect(emailInput).toHaveAttribute("required");
  });

  it("the password input should be a required fields", () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const passwordInput = screen.getByPlaceholderText("Password");
    expect(passwordInput).toBeInTheDocument();
    expect(passwordInput).toHaveAttribute("required");
  });

  it("on initial render, the submit button is enabled", () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const submitButton = screen.getByRole("button", { name: "Login" });
    expect(submitButton).not.toBeDisabled();
  });
});

//test login functionality
describe("Test Login Functionality", () => {
  it("should show error message on password and email when the form is submitted with empty fields", async () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(emailInput, { target: { value: "" } });
    fireEvent.change(passwordInput, { target: { value: "" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toHaveTextContent("Email is a required field"));
    await waitFor(() => expect(screen.queryByTestId("password-error-message")).toHaveTextContent("Password is a required field"));
  });

  it("should show error message email when the form is submitted with empty email", async () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(emailInput, { target: { value: "" } });
    fireEvent.change(passwordInput, { target: { value: "Digicel.1" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toHaveTextContent("Email is a required field"));
  });

  it("should show error message password when the form is submitted with empty password", async () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(emailInput, { target: { value: "test@test.com" } });
    fireEvent.change(passwordInput, { target: { value: "" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("password-error-message")).toHaveTextContent("Password is a required field"));
  });

  it("should show error message email if email missing @", async () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(emailInput, { target: { value: "testtest.com" } });
    fireEvent.change(passwordInput, { target: { value: "" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toHaveTextContent("Email must be a valid email"));
  });

  it("should show error message email if email missing .", async () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(emailInput, { target: { value: "test@testcom" } });
    fireEvent.change(passwordInput, { target: { value: "" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toHaveTextContent("Email must be a valid email"));
  });

  it("should show error message password if password is least than 8 characters .", async () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(emailInput, { target: { value: "test@test.com" } });
    fireEvent.change(passwordInput, { target: { value: "Digi" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("password-error-message")).toHaveTextContent("Password must be at least 8 characters"));
  });

  it("shouldn't display error", async () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(emailInput, { target: { value: "test@test.com" } });
    fireEvent.change(passwordInput, { target: { value: "Digicel.1" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("password-error-message")).toBeNull());
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toBeNull());
  });
});
