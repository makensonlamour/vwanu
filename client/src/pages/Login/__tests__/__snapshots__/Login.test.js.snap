// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Dom Elements match snapshot 1`] = `
<div
  className=""
>
  <div
    className="lg:grid lg:grid-cols-2 lg:gap-4"
  >
    <div
      className="hidden lg:justify-center lg:block"
    >
      <div
        className="bg-blue-600 rounded-t-3xl h-[90%] m-6"
      >
        <p
          className="text-center text-white text-2xl py-5"
        >
          <img
            alt="logo_vwanu"
            className="w-1/3 m-auto"
            src="Asset_3.png"
          />
        </p>
        <p
          className="text-5xl text-yellow-500 font-bold py-8 text-center"
        >
          Welcome
        </p>
        <p
          className="text-white text-xl font-semibold text-center pb-2"
        >
          Share your Voice and Change
        </p>
        <p
          className="text-white text-xl font-semibold text-center"
        >
          The Haitian Community!
        </p>
      </div>
      <div
        className="bg-yellow-400 rounded-b-3xl rounded-tr-[130px] h-[50%] -mt-[9rem] m-6"
      />
    </div>
    <div
      className="place-items-center mx-10 md:mx-20"
    >
      <div
        className="mt-4 mb-6 lg:hidden"
      >
        <img
          alt="logo_vwanu"
          className="w-2/5 m-auto"
          src="Asset_2.png"
        />
      </div>
      <div
        className="place-content-center my-8"
      >
        <p
          className="text-center mt-4"
        >
          <span
            className="text-md text-orange-500 font-semibold md:text-2xl"
          >
            Not a member ?
          </span>
           
          <a
            className="btn btn-primary btn-sm rounded-full text-base-200 normal-case md:px-8 ml-1"
            data-testid="registerBtn"
            href="/register"
            onClick={[Function]}
          >
            Register
          </a>
        </p>
      </div>
      <div
        className="px-10 pt-5 pb-10 card shadow-2xl rounded-3xl"
        role="form"
      >
        <h1
          className="card-title text-orange-500 font-bold text-2xl"
        >
          Sign in to Vwanu
        </h1>
        <div
          className="form-control"
        >
          <label
            className="label"
          >
            <span
              className="label-text"
            />
          </label>
          <input
            autoCapitalize="none"
            autoComplete="new-email"
            className="input bg-blue-200 text-blue-600 font-semibold rounded-full input-primary border-none lg:input-lg"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Email"
            required={true}
            type="email"
            value=""
          />
        </div>
        <div
          className="form-control"
        >
          <label
            className="label"
          >
            <span
              className="label-text"
            />
          </label>
          <input
            autoCapitalize="none"
            autoComplete="new-email"
            autoCorrect="false"
            className="input mt-2 bg-blue-200 text-blue-600 font-semibold rounded-full input-primary border-none md:px-6 lg:input-lg"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            value=""
          />
        </div>
        <div
          className="form-control mt-3"
        >
          <label
            className="label"
          >
            <input
              className="checkbox"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="label-text ml-2 text-sm lg:text-lg lg:ml-0"
            >
              Remember me
            </span>
            <span
              className="ml-8 text-sm lg:text-lg"
            >
              <a
                className="text-orange-500 font-bold"
                href="/forgot-password"
                onClick={[Function]}
              >
                Forgot password
              </a>
            </span>
          </label>
        </div>
        <button
          className="btn btn-primary mt-4 normal-case rounded-full text-base-200 text-md btn-sm lg:btn-md"
          data-testid="loginBtn"
          onClick={[Function]}
          type="submit"
        >
          Login
        </button>
        <div
          className="mt-9 text-center"
        >
          <span
            className="text-blue-600 text-center inline text-md md:text-lg"
          >
             Or sign in with 
            <a
              className="text-blue-500"
              href="/"
              onClick={[Function]}
            >
              <svg
                className="text-2xl inline mx-1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 8.049c0-4.446-3.582-8.05-8-8.05C3.58 0-.002 3.603-.002 8.05c0 4.017 2.926 7.347 6.75 7.951v-5.625h-2.03V8.05H6.75V6.275c0-2.017 1.195-3.131 3.022-3.131.876 0 1.791.157 1.791.157v1.98h-1.009c-.993 0-1.303.621-1.303 1.258v1.51h2.218l-.354 2.326H9.25V16c3.824-.604 6.75-3.934 6.75-7.951z"
                />
              </svg>
            </a>
            <a
              className="text-sky-600"
              href="/"
              onClick={[Function]}
            >
              <svg
                className="text-2xl inline mx-1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.026 15c6.038 0 9.341-5.003 9.341-9.334 0-.14 0-.282-.006-.422A6.685 6.685 0 0 0 16 3.542a6.658 6.658 0 0 1-1.889.518 3.301 3.301 0 0 0 1.447-1.817 6.533 6.533 0 0 1-2.087.793A3.286 3.286 0 0 0 7.875 6.03a9.325 9.325 0 0 1-6.767-3.429 3.289 3.289 0 0 0 1.018 4.382A3.323 3.323 0 0 1 .64 6.575v.045a3.288 3.288 0 0 0 2.632 3.218 3.203 3.203 0 0 1-.865.115 3.23 3.23 0 0 1-.614-.057 3.283 3.283 0 0 0 3.067 2.277A6.588 6.588 0 0 1 .78 13.58a6.32 6.32 0 0 1-.78-.045A9.344 9.344 0 0 0 5.026 15z"
                />
              </svg>
            </a>
            <a
              className="text-red-600"
              href="/"
              onClick={[Function]}
            >
              <svg
                className="text-2xl inline mx-1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256,8C119.1,8,8,119.1,8,256S119.1,504,256,504,504,392.9,504,256,392.9,8,256,8ZM185.3,380a124,124,0,0,1,0-248c31.3,0,60.1,11,83,32.3l-33.6,32.6c-13.2-12.9-31.3-19.1-49.4-19.1-42.9,0-77.2,35.5-77.2,78.1S142.3,334,185.3,334c32.6,0,64.9-19.1,70.1-53.3H185.3V238.1H302.2a109.2,109.2,0,0,1,1.9,20.7c0,70.8-47.5,121.2-118.8,121.2ZM415.5,273.8v35.5H380V273.8H344.5V238.3H380V202.8h35.5v35.5h35.2v35.5Z"
                />
              </svg>
            </a>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
